# 4.4-D1 Data Wrangling Procedures

In [2]:
#Dropping columns in a dataframe.
import pandas as pd
import numpy as np
import os  

df_ords = pd.read_csv(r'/Users/katep/08-2023 Instacart Basket Analysis/02 Data/Original Data/orders.csv', index_col=False)
df_ords2= df_ords.drop(columns=['eval_set'])
df_ords2['days_since_prior_order'].value_counts(dropna=False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [33]:
#Renaming Columns
df_ords2.rename(columns={'order_dow':'orders_day_of_week'}, inplace=True)
df_ords2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [35]:
#Changing a variable's data type
df_ords2['order_id']= df_ords2['order_id'].astype('str')
df_ords2['order_id'].dtype

dtype('O')

In [27]:
#Transposing Data
path= r'/Users/katep/08-2023 Instacart Basket Analysis'
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col=False)
df_dep_t= df_dep.T
df_dep_t.reset_index()


index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [28]:
new_header = df_dep_t.iloc[0]
new_header


0    department
Name: department_id, dtype: object

In [31]:
df_dep_t_new= df_dep_t[1:]
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [41]:
#set the header row as the dataframe header
df_dep_t_new.columns = new_header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# D2 starts

In [24]:
#D2. Find another identifier variable in the df_ords dataframe and change it to a suitable format.
df_ords2['user_id']= df_ords2['user_id'].astype('str')
df_ords2['user_id'].dtype

dtype('O')

In [39]:
#D3. Look for a variable in df_ords with an unintuitive name and change its name without overwriting the dataframe.
df_ords2.rename(columns={'order_number':'order_sequence'}, inplace=True)
df_ords2.head()

order_id  user_id  order_sequence  orders_day_of_week  order_hour_of_day  \
0  2539329        1               1                   2                  8   
1  2398795        1               2                   3                  7   
2   473747        1               3                   3                 12   
3  2254736        1               4                   4                  7   
4   431534        1               5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [19]:
#D4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.
df_ords2['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

D4-2. 10AM has the most orders coming in. Around and after midnight, orders drop gradually and there are the least orders.

In [29]:
#D5.Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.
data_dict= df_dep_t_new.to_dict('index')
print(data_dict.get('4'))

{'department': 'produce'}


In [5]:
#D6.The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)
df_breakfast_prods = df_prods[df_prods['department_id']==14]
df_breakfast_prods

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [8]:
#D7. Products that customers might use to throw dinner parties. Include items from the following departments: alcohol, deli, beverages, and meat/seafood.
df_party_prods = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]
df_party_prods

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

#D8. How many rows does the last dataframe you created have?
    7650 rows.

In [13]:
#D9. Extract all the information about the customer with a "user_id" of “1.” 
df_customer=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col=False)
df_customer_Id1= df_customer[df_customer['user_id']==1]
df_customer_Id1

user_id First Name  Surnam  Gender    STATE  Age date_joined  \
134862        1      Linda  Nguyen  Female  Alabama   31   2/17/2019   

        n_dependants fam_status  income  
134862             3    married   40423

In [20]:
#D9-2. 
df_customer1_order= df_ords2[df_ords2['user_id']==1]
df_customer1_order

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   
10  1187899        1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

#D10. You also need to provide some details about this user’s behavior. What basic stats can you provide 
      based on the information you have?
      This user made orders in the office hours during week days. The first time this user made an order was 2 weeks
      before her second order and then she started to make orders more frequently.

In [40]:
#D12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.
df_ords2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))


In [42]:
#D13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder 
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))